### PR TITLE
Adding support for hooks/decorators by using FastifyRequest for websocket handlers

### LINKED ISSUE #70: Start using FastifyRequest objects for the upgrade request so hooks fire and decorators can be used
## 🚀 Feature Proposal

`fastify-websocket` currently passes the raw `http.IncomingMessage` into websocket route handlers, not a `FastifyRequest` object. I think instead passing a `FastifyRequest` to handlers would improve a number of things.

## Motivation

First off, it is unfortunate to not have access to `FastifyRequest`'s basic helpers. The higher level object is more useful than the raw `IncomingMessage` and has some handy doodads for getting say route params or the logger to work with. Second, fastify promotes and has a really healthy ecosystem of things that decorate `FastifyRequest` that websocket handlers also can't access. Things like a session from `fastify-secure-session`, utilities from `fastify-url-data`, or user-land decorators to get a database client or what have you aren't available in route handlers.

Finally, it's unfortunate that hooks don't run for websocket handlers. We'd have to discuss semantically how this might work, but, if we could come up with sane semantics for this, it'd be really handy. I want to use hooks to implement authorization for example -- deny requests based on the presence of a session value or whatever. This same logic needs to apply to work done over normal HTTP requests as well as work done over websocket connections, but currently I have to duplicate it in a hook and in a `verifyClient` option. One place (especially one place that had access to decorators) would be simpler.

I also think there's an argument to be made that websocket handlers being different like this violates the principle of least surprise.

## How

This'd definitely be a breaking change, so knowing that, we'd either introduce a new API or just do a major version bump. With that, I think it'd be possible to actually implement this by using `ws` ability to bind to an existing `http.Server`, as is detailed here: https://www.npmjs.com/package/ws#external-https-server

I think for an `UPGRADE` request we'd probably not want to provide a `FastifyReply` object, or we'd want to change the API to use `FastifyReply` to actually start the websocket connection. I think there's really only one thing you want to do with an upgrade request, which is delegate it to the websocket server, so I think having to do that manually isn't really worth it, and that it's a special enough scenario that having a different handler signature is actually a good thing anyways. So I'd suggest having websocket handlers look pretty similar to how the do right now.

Hooks wise, the semantics that make sense to me are firing everything up until the reply is sending. So, `onRequest`, `preParsing`, `preValidation`, `preHandler`, `preSerialization`, and `onError` could all be relied on, but `onSend` and `onResponse` would never be called and would be documented as such. I am not really familiar with Fastify's internals enough to say how hard it would be to actually do this weird half-chain of hooks, but were that easy enough, this makes sense to me.

## Context

I don't really know why `fastify-websocket` works the way it does right now, but if I had to guess, I'd say it's because its really simple to implement. That's not a bad reason. I am finding that I have to do more work in userland because of this though, and I think implementing this inside the library instead of in bespoke and crappy ways outside it each time would save devs lots of time.

Thoughts?

### PR DESCRIPTION
**Pull request by AyoubElk**
This is about [fastify/fastify-websocket/#70](https://github.com/fastify/fastify-websocket/issues/70), it closes #70 and is dependent on the changes here [fastify/fastify/#2733](https://github.com/fastify/fastify/pull/2733)

#### This:
- removes the internal find-my-way router used for websocket routes
- dispatches websocket incoming connections through the normal fastify request cycle
- runs all hooks up to handler execution
- allows access to FastifyRequest object inside websocket handlers

#### Implementation details:
- In order to run the hooks and create the necessary internals objects `FastifyRequest`, `FastifyReply`, and `Context`, the incoming connection needs to be dispatched to the fastify instance router. This was done by updating the fastify repository in order to expose the routing utility (precisely the `httpHandler` function in `fastify.js`)
- In the `onRoute` hook, a wrapper-handler is created and an if-else statement is used to allow checking if we're handling a normal http request or websocket connection, and either run `handler` or `wsHandler`
- In order to only run the hooks up to handler execution, I followed @airhorns  suggestion and added a `Reply.hijack()` method to `FastifyReply` which marks it as hijacked, this allows plugins to exit the `Reply.send()` method and skip the execution of remaining hooks
- The global handler is still supported, however it receives as a param a req object of type `IncomingMessage`, this is because no match was find and we use the `defaultRoute` function instead of the `routeHandler` which creates the `FastifyRequest`, `FastifyReply`, and `Context` objects. Also, in order to support this global handler, we needed a way to override the fastify `defaultRoute` function and handle websocket connections as well instead of defaulting to `FourOhFour` router lookup. This was done by exposing it from the fastify instance and monkeypatching it inside the plugin.

Looking forward to your feedback!
Thanks

#### Checklist
- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.d.ts
Comment (airhorns) [inline index.d.ts:43]: 😍

File: index.js
Comment (airhorns) [inline index.js:55]: Could you think of a reason why `hijack` would ever need to be async or return something? I can't off the top of my head, but this style of API that is super side-effect-ful always makes me a bit nervous.

File: index.js
Comment (airhorns) [inline index.js:86]: I think this is a great solution to the global handler thing. We could consider dropping support for global handlers IMO, I am not really sure when you would want to effectively ignore the path, especially knowing that this library is for integrating with fastify, which definitely does not ignore the path for normal HTTP requests. Now that websockets would be handled all smartly using the router we could tell folks to migrate to using a route like `/*` if they really did want a global handler. I will leave it to the maintainers to make that call though.

Comment (airhorns) [reply index.js:86]: Does this monkeypatch change the semantics of when the global handler is fired? If I understand correctly, before this change, a global websocket handler would handle any websocket request not matched by the internal `fastify-websocket` router. Now, since this is going through the main fastify instance router, if a websockets request matched a route within the main router, fastify would run that handler and not this default one, right?

File: test/router.js
Comment (airhorns) [inline test/router.js:544]: Would be good to add a test for both a websocket route and a non websocket route that are registered on the exact same path, and make sure both handlers can be invoked depending on the headers? I think that is a use case the HTTP spec supports. If it's too complicated to support we should probably warn or throw as well.

File: test/types/index.test-d.ts
Comment (airhorns) [inline test/types/index.test-d.ts:65]: Sweet, sweet bliss

**Review by airhorns — COMMENTED** [review]
Lookin super duper awesome @AyoubElk , great work!

File: index.js
Comment (AyoubElk) [reply index.js:86]: You're right, I haven't considered this use case. I'll think of some way to handle it.
Personally, I'm all for dropping support for the global handler if the maintainers feel that would be the best way forward.

**Comment by mcollina** [comment]
Good work!

File: index.js
Comment (airhorns) [reply index.js:86]: @mcollina would you mind weighing in specifically on if you think keeping the global handler functionality this library previously exposed is worth it? By global, I mean "routeless" matching, where the same websockets handler would be invoked regardless of the initial request path. The alternative this PR enables would be "routeful" matching, where you have to register a route to get websocket connections, but you could use routes like `/*`if you wanted the same-ish behaviour. This PR tries to keep around the routeless matching, but it causes a bunch of the complexity around the `defaultRoute` exposure and whatnot that might be nice to drop. I think "routeful" matching makes a lot more sense and is a lot more explicit, and if I had to guess the global handler thing exists just cause it was easy and how the `ws` library tends to work.

Comment (AyoubElk) [reply index.js:86]: @mcollina To give more context regarding the issue I'm facing, the global handler is currently implemented by updating the `defaultRoute` function (_which I'll be changing by updating `find-my-way` as well to add a `setDefaultRoute` method and avoid the monkeypatching as suggested_), however this doesn't support the edge case where the `defaultRoute` function won't be called if the router might already have a route registered with the path the websocket connection is using.
To solve this, the solution I'm currently considering is to check in the `onRoute` hook if the route is a wildcard route with `GET` method, in which case I can do something like this:
```js
routeOptions.handler = (request, reply) => {
  if (request.raw[kWs]) {
    reply.hijack()
    let result
    if (isWildcardRoute) {
      result = handle.call(fastify, request.raw[kWs], request.raw)
    } else {
      result = wsHandler.call(fastify, request.raw[kWs], request)
    }
    if (result && typeof result.catch === 'function') {
      result.catch(err => request.raw[kWs].destroy(err))
    }
  } else {
    return handler.call(fastify, request, reply)
  }
}
```
However, this solution isn't complete because it's possible the user has not registered a wildcard route. I might try registering one in the plugin but there is no way to catch and ignore a duplicate route error as far as I can tell, and hooks are executed after all routes are registered so I can't register it after making sure it doesn't exist already.

Comment (mcollina) [reply index.js:86]: > however this doesn't support the edge case where the defaultRoute function won't be called if the router might already have a route registered with the path the websocket connection is using.
To solve this, the solution I'm currently considering is to check in the onRoute hook if the route is a wildcard route with GET method, in which case I can do something like this:

I don't understand this. Anyway, the solution looks ok, I'll see when there is a test.

**Review by fox1t — APPROVED** [review]
Amazing!

File: test/router.js
Comment (airhorns) [reply test/router.js:544]: Also would be good to add a test showing when the fastify server error handler is invoked. I don't think it's really possible to invoke it after the request has been fully upgraded, but, it might make sense for it to still be invoked before a request is hijacked. Same as the `preValidation` handlers or what have you, it'd be good if websocket-y requests had their errors handled the same way as other fastify routes before they actually become websocket-y.

**Comment by AyoubElk** [comment]
As an update, I'm still adding relevant tests to both repositories.
Like requested, in here I have to add tests to cover when the remaining hooks 'onError', and 'onSend' will be called (depending on if the request was upgraded successfully and hijacked)
In `fastify`, I have to make some updates to make sure `onError` is executed even if the reply is hijacked (because it is executed inside `reply.send`) and other more tests.

File: test/router.js
Comment (AyoubElk) [inline test/router.js:240]: @mcollina, this test normally should fail (I changed it to pass just so I can push)
We already have an http route registered with the same path, so the router then matches it and returns the http handler which has no information about the websocket connection.

This is the issue I was referring to the other day caused by having the global handler operate the way it does right now.

**Comment by airhorns** [comment]
> In fastify, I have to make some updates to make sure onError is executed even if the reply is hijacked (because right now the hook is executed inside reply.send after the hijacking is done, so it's skipped) and add tests to cover the changes.

I think that these semantics will be really hard to get right and that we should consider making the rule "no error handling willy apply after a request is hijacked". The reason for me would be that Fastify can't really know what's going to happen after hijacking, and so it will have a hard time wrapping all the code with something that would catch errors. Even if it could wrap some things, I don't think it could wrap everything, so you'd end up with some errors being handled by fastify and some not, and end up still having to build your own error handling, but spend more time figuring out which ones fastify handled and which ones it didn't. In the case of websockets, I think it'd be extra hard for fastify to catch errors inside websocket message handlers, and extra hard to know how to handle them, and so it wouldn't be able to catch those, but once developers saw it handled some errors after hijacking, they might expect Fastify to handle errors from other strange asynchronous stacks too. 

I think `onSend` should just never be called also.

File: test/router.js
Comment (AyoubElk) [reply test/router.js:240]: An interesting solution to keep the global handler here would be to use the `onRoute` hook to wrap all handlers (not just the ones with websocket handlers defined) and make them call the global handler in case it's a websocket connection and they have no websocket handler defined.

```js
routeOptions.handler = (request, reply) => {
    if (request.raw[kWs]) {
        reply.hijack()
        // wsHandler can be either the route handler or global handler
        const result = wsHandler.call(fastify, request.raw[kWs], request)
        if (result && typeof result.catch === 'function') {
            result.catch(err => request.raw[kWs].destroy(err))
        }
    } else {
        return handler.call(fastify, request, reply)
    }
}
```

With this and the `defaultRoute` updated, we can cover all use cases, the downside would be the overhead an additional function call and condition check for each handler call, but I think that's worth it if we want to keep the global handler. 

What do you think @mcollina?

Comment (AyoubElk) [reply test/router.js]: I implemented this solution, it works if the user registers all routes after the `fastify-websocket` plugin is registered, otherwise the `onRoute` hook won't be called for the previous routes and the handler can't be updated to call the global handler in case of a match.

Also, the global handler is called in 2 scenarios:
1/ A route with no websocket handler is matched, the global handler is called instead.
2/ No route match is found, the updated `defaultRoute` is called

This means that for `1` the hooks are executed and we go through the normal fastify request cycle before running the global handler, but for `2` we don't.

I've tried hard to find a solution that would allow us to keep the global handler and cover all use cases, but this is the best one I was able to come up with, if it's acceptable to document this behavior in the Readme file we'll be able to keep the global handler, otherwise we should discuss alternatives.

Let me know what you think

**Comment by AyoubElk** [comment]
Not sure why the tests are timing out here btw, everything runs fine for me.

**Review by mcollina — COMMENTED** [review]
Good work! This needs some docs update!

File: test/router.js
Comment (AyoubElk) [reply test/router.js]: @mcollina what do you think about this? 
I've updated the docs as well

**Comment by mcollina** [comment]
It seems node 10 is failing.

**Comment by AyoubElk** [comment]
> It seems node 10 is failing.

It seems that this is caused by `t.expectUncaughException` not receiving anything and timing out for this node version, I've been trying to figure out the cause but so far haven't been able to do so.

Any advice as to what might be the cause would be much appreciated.

File: test/hooks.js
Comment (mcollina) [inline test/hooks.js:101]: Why is `expectUncaughtException` used here?

File: test/hooks.js
Comment (mcollina) [inline test/hooks.js:126]: Why is `expectUncaughtException` used here?

File: test/hooks.js
Comment (mcollina) [inline test/hooks.js:95]: `expectUncaughtException` accepts a function but you are passing an object.

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't really understand why you are using `expectUncaughtException` at all. You should not really need to use it. I also think the behavior is wrong, as we should never reach uncaughtException in these two tests.

File: test/hooks.js
Comment (AyoubElk) [reply test/hooks.js:101]: I've used it because when calling `conn.destroy(err)`, this is thrown:
```
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: uncaught
    at Object.<anonymous> (/Users/ayoub//fastify-websocket/exampl.js:8:27)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
Emitted 'error' event at:
    at Duplex.duplexOnError (/Users/ayoub/fastify-websocket/node_modules/ws/lib/stream.js:37:10)
    at Duplex.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Using `expectUncaughtException` worked to intercept this for the other node versions that pass the tests.

File: test/router.js
Comment (mcollina) [inline test/router.js:5]: why is this commented out?

File: index.js
Comment (mcollina) [inline index.js:119]: I think we should log the error somewhere.

File: index.js
Comment (mcollina) [inline index.js:113]: I think the error handler should also include the fastify objects.

I would also use the errorback pattern: `err, conn, req, reply` for example.

**Review by mcollina — COMMENTED** [review]
amazing work, we are almost there.

File: index.js
Comment (AyoubElk) [reply index.js:119]: Do you think using `fastify.log` is the right way to do it?

File: test/router.js
Comment (AyoubElk) [reply test/router.js:5]: Forgot to remove it after I was done with it.

File: index.js
Comment (AyoubElk) [reply index.js:113]: The error handler is used for errors of both the global and per-route handlers, so it can't take `FastifyRequest`/`FastifyReply` objects.

Would you want it to receive core nodeJS req/res objects?

File: index.js
Comment (mcollina) [reply index.js:119]: You should really use `req.log` as it already has some details on the current requests. You might also want to attach that to the conn.

File: index.js
Comment (mcollina) [reply index.js:113]: no worries then!

File: index.js
Comment (AyoubElk) [reply index.js:119]: Since the error handler doesn't _always_ have access to the `FastifyRequest`, there would be no way to use `req.log`

Comment (mcollina) [reply index.js:119]: Use fastify.log for now, but we should add a few TODOs / issues about logging.

Comment (AyoubElk) [reply index.js:121]: Done, I'd like to note that removing usage of the global handler would make this easier, allow all handlers to have access to `FastifyRequest`/`FastifyReply` objects, and keep things uniform.

Have you considered that possibility?

Comment (mcollina) [reply index.js:121]: Let's remove it. Let's add a test that verifies it all work with a wildcard route: `/*`.

Comment (AyoubElk) [reply index.js:110]: I've removed usage of the global handler (keeping `noHandle` to close connections on unregistered routes)

File: README.md
Comment (mcollina) [inline README.md:169]: Does the path option still make sense? We already have the router.

File: index.js
Comment (mcollina) [inline index.js:36]: I think this should be called with error as an argument in case it's not a websocket.

File: index.js
Comment (mcollina) [inline index.js:99]: We should log something.

File: test/base.js
Comment (mcollina) [inline test/base.js:80]: Can you please add a normal /* route and see if everything keep working fine?

File: README.md
Comment (AyoubElk) [reply README.md:169]: You're right, it doesn't.
I'll update the readme file to advise against using the `path` option.
I also think it would be best to delete it (if provided) from the `options` object before passing it to the websocket server, while logging something to warn the user that it shouldn't be used. 

Let me know if you think it should be handled differently.

File: index.js
Comment (AyoubElk) [reply index.js:36]: Do you mean passing the error to the `done` function?

Quote [from](https://www.fastify.io/docs/master/Hooks/#onerror):

> Notice: unlike the other hooks, pass an error to the done function is not supported.

File: README.md
Comment (mcollina) [reply README.md:169]: Logging something if this is passed through is ok. Let's drop it from the docs as well.

File: test/base.js
Comment (mcollina) [reply test/base.js:81]: I meant having a test with both `/*` websocket and `/*` normal. I just want to triple-check this as wildcards are very common (for example in fastify-static).

File: index.js
Comment (mcollina) [reply index.js:39]: You are right, forget about it!

**Comment by AyoubElk** [comment]
> Good work, only the test is missing!

I've added a test with both normal amd websocket wildcard routes already, is that what you're referring to?

**Comment by mcollina** [comment]
> > Good work, only the test is missing!
> 
> I've added a test with both normal amd websocket wildcard routes already, is that what you're referring to?

I can't find it. Something with

```js
app.get('/*', { websocket: true }, ...)
app.get('/*, ...)
```

In the same server.

**Comment by AyoubElk** [comment]
> > > Good work, only the test is missing!
> > 
> > I've added a test with both normal amd websocket wildcard routes already, is that what you're referring to?
> 
> I can't find it. Something with
> 
> ```js
> app.get('/*', { websocket: true }, ...)
> app.get('/*, ...)
> ```
> 
> In the same server.

https://github.com/fastify/fastify-websocket/commit/8dc9038aeeaf2d3adbd252ad519f486669d73353

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by AyoubElk** [comment]
Resolved all conflicts, all that's left here is to update Fastify's version once the changes are published.

**Comment by mcollina** [comment]
I just shipped Fastify v3.11.0 with the required change. Be sure to set it as the minimum version when calling fastify-plugin.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by airhorns** [comment]
Are we waiting on review from other folks?

**Comment by mcollina** [comment]
Ideally yes. It's a fundamental change.

**Review by zekth — APPROVED** [review]
lgtm good work!

**Review by DRoet — APPROVED** [review]
Amazing work!

**Pull request merged**
